src/main/flight/pid.c

```diff
@@ -343,8 +364,50 @@ static float calcWingTpaArgument(void)
 
     return tpaArgument;
 }
+
+static void updateStermTpaFactor(int axis, float tpaFactor)
+{
+    float tpaFactorSterm = tpaFactor;
+    if (pidRuntime.tpaCurveType == TPA_CURVE_HYPERBOLIC) {
+        const float maxSterm = tpaFactorSterm * (float)currentPidProfile->pid[axis].S * S_TERM_SCALE;
+        if (maxSterm > 1.0f) {
+            tpaFactorSterm *=  1.0f / maxSterm;
+        }
+    }
+    pidRuntime.tpaFactorSterm[axis] = tpaFactorSterm;
+}
+
+static void updateStermTpaFactors(void) {
+    for (int i = 0; i < XYZ_AXIS_COUNT; i++) {
+        float tpaFactor = pidRuntime.tpaFactor;
+        if (i == FD_YAW && currentPidProfile->yaw_type == YAW_TYPE_DIFF_THRUST) {
+            tpaFactor = pidRuntime.tpaFactorYaw;
+        }
+        updateStermTpaFactor(i, tpaFactor);
+    }
+}
 #endif // USE_WING
 
+static float wingAdjustSetpoint(float currentPidSetpoint, int axis)
+{
+#ifdef USE_WING
+    float adjustedSetpoint = currentPidSetpoint;
+    if (!IS_AXIS_IN_ANGLE_MODE(axis)) {
+        const bool skipYaw = axis == FD_YAW && currentPidProfile->yaw_type == YAW_TYPE_DIFF_THRUST;
+        if (pidRuntime.tpaFactorSterm[axis] > 0.0f && pidRuntime.tpaFactor > 0.0f && !skipYaw) {
+            adjustedSetpoint = currentPidSetpoint * pidRuntime.tpaFactorSterm[axis] / pidRuntime.tpaFactor;
+        }
+    }
+
+    DEBUG_SET(DEBUG_WING_SETPOINT, 2 * axis, lrintf(currentPidSetpoint));
+    DEBUG_SET(DEBUG_WING_SETPOINT, 2 * axis + 1, lrintf(adjustedSetpoint));
```

Comment:
@limon need some heads up here. How to interpret the index here?
```
2 * xyz = { 2, 4, 6 }
2 * xyz + 1 = { 3, 5, 7 }
```

Think should be 
```
xyz = { 0, 1, 2 }
xyz + 3 = { 3, 4, 5 }
```

Reply:
@haslinghuis:: axis is zero based, so 0,2,4 and 1,3,5